js/src/dom/data.js

```diff
@@ -11,57 +11,45 @@
  * ------------------------------------------------------------------------
  */
 
-const mapData = (() => {
-  const storeData = {}
-  let id = 1
-  return {
-    set(element, key, data) {
-      if (typeof element.bsKey === 'undefined') {
-        element.bsKey = {
-          key,
-          id
-        }
-        id++
-      }
+const elementMap = new WeakMap()
```

Comment:
Used a WeakMap here. The advantage is, it holds weak references on the elements and thus shouldn't prevent them from getting garbage collected. Not sure if there are any concerns about this.

Reply:
Not super familiar with this either. I'd go with a Map personally, but maybe wait for someone else to chime in

Reply:
Agree with @alpadev here

> But `WeakMap` automatically also removes the key along with the information kept by that key when the key (as object) becomes unreachable.

So probably in the future when supporting multiple plugin instances it should be revisited because `dispose` method in the plugins sets the element to null. What do you think? 🙂

Reply:
@rohit2sharma95 tbh I'm not completely aware how WeakMaps internally work, but to my understanding, using an object (or element) as a key here doesn't alter the general behaviour of the garbage collection. It rather uses a weak reference to an object that can be freed automatically when the object has no other reference, instead of a strong reference that would prevent an object from being collected until that reference is explicitely removed.

In that case you mention here, when there are multiple plugin instances each referencing the same element. Setting the element to null in one instance (e.g. `this._element = null`) shouldn't have an impact on any other reference to the same element or the element itself. It's not the deleting the element but rather overwriting that single reference to it with null. 

As such the key-element of the WeakMap should stay intact, as this isn't altered by the code.

Reply:
I'm still hesitant to using a WeakMap here. I'd rather err on the safe side and go with a Map personally.